docs/extensions/api.md

```diff
@@ -2,425 +2,563 @@ title:      Extensions API
+parser. This is an excellent place to clean up bad characters or to extract 
+portions for later processing that the parser may otherwise choke on.
+
+Preprocessors inherit from `markdown.preprocessors.Preprocessor` and
+implement a `run` method, which takes a single parameter `lines`. This parameter is
+the entire source text stored as a list of Unicode string, one per line.  `run` should
+return its processed list of of Unicode strings, one per line.
+
+#### Example
+
+This simple example removes any lines with 'NO RENDER' before processing:
 
 ```python
 from markdown.preprocessors import Preprocessor
+import re
 
-class MyPreprocessor(Preprocessor):
+class NoRender(Preprocessor):
+    """ Skip any line with words 'NO RENDER' in it. """
     def run(self, lines):
         new_lines = []
         for line in lines:
-            m = MYREGEX.match(line)
+            m = re.search("NO RENDER", line)
             if m:
-                # do stuff
+                pass   # skipping this line
             else:
                 new_lines.append(line)
         return new_lines
 ```
 
-## Inline Patterns {: #inlinepatterns }
+#### Usages
+
+Some preprocessors in the Markdown source tree include:
+
+| Class  | Kind   | Description |
+| ----------------------------|-----------|-------------------------------------------------
+| [`NormalizeWhiteSpace`][c1]   | built-in  | Normalizes whitespace by expanding tabs, fixing `\r` line endings, etc. |
+| [`HtmlBlockPreprocessor`][c2] | built-in  | Removes html blocks from the text and stores them for later processing |
+| [`ReferencePreprocessor`][c3] | built-in  |Removes reference definitions from text and stores for later processing |
+| [`MetaPreprocessor`][c4]      | extension |Strips and records meta data at top of documents |
+| [`FootnotesPreprocessor`][c5] | extension | Removes footnote blocks from the text and stores them for later processing |
+
+[c1]: https://github.com/Python-Markdown/markdown/blob/master/markdown/preprocessors.py#L62
+[c2]: https://github.com/Python-Markdown/markdown/blob/master/markdown/preprocessors.py#L74
+[c3]: https://github.com/Python-Markdown/markdown/blob/master/markdown/preprocessors.py#L339
+[c4]: https://github.com/Python-Markdown/markdown/blob/master/markdown/extensions/meta.py#L32
+[c5]: https://github.com/Python-Markdown/markdown/blob/master/markdown/extensions/footnotes.py#L205
+
+### Block Processors {: #blockprocessors }
+
+A block processor parses blocks of text and adds new elements to the `ElementTree`. 
+Blocks of text, separated from other text by blank lines, may have a different syntax 
+and produce a differently structured tree than
+other Markdown.  Block processors excel at code formatting, 
+equation layouts, and tables.  
+
+Block processors inherit from `markdown.blockprocessors.BlockProcessor`, 
+are passed `md.parser` on initialization,
+and implement both the `test` and `run` methods:  
```

Comment:
We have some weird line wrapping going on here.

Reply:
Actually, there appears to be less-than-ideal wrapping throughout the document. It would be nice to have that all cleaned up.

Reply:
I'll try to clean some up; it's a manual process PyCharm.  The ragged edge is only visible to those editing the document and will not show when the documentation is rendered.  Be forgiving if some stays ragged. :)

Reply:
Our [Contributing Guide][2] is pretty clear that we want neat code and documentation. And yes, that means the source documentation. As the creator of Markdown [explains][1]:

> The overriding design goal for Markdown’s formatting syntax is to make it as readable as possible. The idea is that a Markdown-formatted document should be publishable as-is, as plain text...

Ragged text is not "as readable as possible." Always assume the document is being read as raw Markdown text.

That said, sometimes it is not possible due to a long URL or some other formatting. And yes, there have been a few times we let something messy slip in. However, in most cases, a good text editor should be able to clean up hard line breaks with a single command. 

[1]: https://daringfireball.net/projects/markdown/
[2]: https://python-markdown.github.io/contributing/